[ONNX] Assign ONNXScopeName during function substituion

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #82040
* __->__ #82039

Previously only traced IR graph stores module typename and variable name
in `scope` in `node`. This change enables such `scope` info for IR graph
generated by torch script.
Torch script produced IR graphs emit nodes for module object and module
method call. This structured graph is flattened in `function_substition`
pass prior to other ONNX conversion passes.
This PR extends `function_substition` pass to record the module typename
and variable name info in `scope`, while inlining the graph.

BowenBao committed:
[ONNX] Assign ONNXScopeName during function substituion

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Assign ONNXScopeName during function substituion"




[ghstack-poisoned]

AllenTiTaiWang commented on torch/csrc/jit/passes/onnx/function_substitution.cpp:
```diff
@@ -2,11 +2,79 @@
+      out += atom;
+    }
+  }
+  std::cout << " out name: " << out << std::endl;
```
I wonder if this is necessary ...

BowenBao replied:
Indeed, let me think about it ...

BowenBao committed:
Update on "[ONNX] Assign ONNXScopeName during function substituion"




[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Assign ONNXScopeName during function substituion"




[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Assign ONNXScopeName during function substituion"


Previously only traced IR graph stores module typename and variable name
in `scope` in `node`. This change enables such `scope` info for IR graph
generated by torch script.
Torch script produced IR graphs emit nodes for module object and module
method call. This structured graph is flattened in `function_substition`
pass prior to other ONNX conversion passes.
This PR extends `function_substition` pass to record the module typename
and variable name info in `scope`, while inlining the graph.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Assign ONNXScopeName during function substituion"


Previously only traced IR graph stores module typename and variable name
in `scope` in `node`. This change enables such `scope` info for IR graph
generated by torch script.
Torch script produced IR graphs emit nodes for module object and module
method call. This structured graph is flattened in `function_substition`
pass prior to other ONNX conversion passes.
This PR extends `function_substition` pass to record the module typename
and variable name info in `scope`, while inlining the graph.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Assign ONNXScopeName during function substituion"


Previously only traced IR graph stores module typename and variable name
in `scope` in `node`. This change enables such `scope` info for IR graph
generated by torch script.
Torch script produced IR graphs emit nodes for module object and module
method call. This structured graph is flattened in `function_substition`
pass prior to other ONNX conversion passes.
This PR extends `function_substition` pass to record the module typename
and variable name info in `scope`, while inlining the graph.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Assign ONNXScopeName during function substituion"


Previously only traced IR graph stores module typename and variable name
in `scope` in `node`. This change enables such `scope` info for IR graph
generated by torch script.
Torch script produced IR graphs emit nodes for module object and module
method call. This structured graph is flattened in `function_substition`
pass prior to other ONNX conversion passes.
This PR extends `function_substition` pass to record the module typename
and variable name info in `scope`, while inlining the graph.

[ghstack-poisoned]

AllenTiTaiWang commented on test/onnx/test_utility_funs.py:
```diff
@@ -1042,6 +1042,21 @@ def forward(self, x, y, z):
+        }
+
+        graph, _, _ = self._model_to_graph(
+            torch.jit.script(model), (x, y, z), input_names=[], dynamic_axes={}
```
Is an empty input_names and dynamic_axes necessary in this cirsumstances? I am not sure.

BowenBao replied:
per discussion, could be improved by revisiting `_model_to_graph` method internals.

AllenTiTaiWang commented on test/onnx/test_utility_funs.py:
```diff
@@ -1042,6 +1042,21 @@ def forward(self, x, y, z):
+            torch.jit.script(model), (x, y, z), input_names=[], dynamic_axes={}
+        )
+        for node in graph.nodes():
+            self.assertIn(node.scopeName(), expected_torch_script_scope_names)
```
nit (I guess): I can see the reason that a hardcoded expected outcomes are used here, just like we have under expect folder. I am concerned that what if there is a change in onnx graph that the nodes are not what we expect now, but the naming logic is still right. In that case, we will have to manually change this hardcoded part here again. But I might be wrong that there shouldn't be this major different in onnx graph?

BowenBao replied:
It's legitimate concern. Hence the test is written with simple nodes such that the nodes rarely change, and test failures are mostly triggered by name discrepancies.

AllenTiTaiWang commented:
cc @justinchuby @shubhambhokare1 @thiagocrepaldi pleas take a look if you guys have time. I think this is a very important new feature.

justinchuby commented:
I assume the PRs in this series are stable & ready for review?

BowenBao committed:
Update on "[ONNX] Assign ONNXScopeName during function substituion"


Previously only traced IR graph stores module typename and variable name
in `scope` in `node`. This change enables such `scope` info for IR graph
generated by torch script.
Torch script produced IR graphs emit nodes for module object and module
method call. This structured graph is flattened in `function_substition`
pass prior to other ONNX conversion passes.
This PR extends `function_substition` pass to record the module typename
and variable name info in `scope`, while inlining the graph.

[ghstack-poisoned]

justinchuby commented on torch/csrc/jit/passes/onnx/function_substitution.cpp:
```diff
@@ -2,11 +2,84 @@
+  }
+  std::string out = "";
+  for (const auto& atom : class_name->atoms()) {
+    bool isInternalTorchAtom = (atom == "__torch__");
```
nit: snake case for var names? https://google.github.io/styleguide/cppguide.html#Variable_Names

BowenBao committed:
Update on "[ONNX] Assign ONNXScopeName during function substituion"


Previously only traced IR graph stores module typename and variable name
in `scope` in `node`. This change enables such `scope` info for IR graph
generated by torch script.
Torch script produced IR graphs emit nodes for module object and module
method call. This structured graph is flattened in `function_substition`
pass prior to other ONNX conversion passes.
This PR extends `function_substition` pass to record the module typename
and variable name info in `scope`, while inlining the graph.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Assign ONNXScopeName during function substituion"


Previously only traced IR graph stores module typename and variable name
in `scope` in `node`. This change enables such `scope` info for IR graph
generated by torch script.
Torch script produced IR graphs emit nodes for module object and module
method call. This structured graph is flattened in `function_substition`
pass prior to other ONNX conversion passes.
This PR extends `function_substition` pass to record the module typename
and variable name info in `scope`, while inlining the graph.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Assign ONNXScopeName during function substituion"


Previously only traced IR graph stores module typename and variable name
in `scope` in `node`. This change enables such `scope` info for IR graph
generated by torch script.
Torch script produced IR graphs emit nodes for module object and module
method call. This structured graph is flattened in `function_substition`
pass prior to other ONNX conversion passes.
This PR extends `function_substition` pass to record the module typename
and variable name info in `scope`, while inlining the graph.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Assign ONNXScopeName during function substituion"


Previously only traced IR graph stores module typename and variable name
in `scope` in `node`. This change enables such `scope` info for IR graph
generated by torch script.
Torch script produced IR graphs emit nodes for module object and module
method call. This structured graph is flattened in `function_substition`
pass prior to other ONNX conversion passes.
This PR extends `function_substition` pass to record the module typename
and variable name info in `scope`, while inlining the graph.

[ghstack-poisoned]

justinchuby commented on test/onnx/test_utility_funs.py:
```diff
@@ -1046,6 +1046,8 @@ def forward(self, x, y, z):
             "test_utility_funs.M::/torch.nn.modules.normalization.LayerNorm::lns.1",
             "test_utility_funs.M::/torch.nn.modules.normalization.LayerNorm::lns.2",
+            "test_utility_funs.M::/test_utility_funs.N::relu/"
+            "torch.nn.modules.activation.ReLU::relu",
```
Is this line too long? maybe just keep it as one line for clarity?

justinchuby commented on torch/csrc/jit/passes/onnx/function_substitution.cpp:
```diff
@@ -60,28 +133,62 @@ void functionCallSubstitution(Block* block) {
+    return graph.current_scope();
+  }
+  if (auto top_module_type = graph.inputs().at(0)->type()->cast<ClassType>()) {
+    auto scope_name = ::torch::jit::onnx::ONNXScopeName::CreateFullScopeName(
```
note: (repeated from another PR) I realized the pytorch project conventionally uses camelCase for function names (instead of PascalCase)

BowenBao committed:
Update on "[ONNX] Assign ONNXScopeName during function substituion"


Previously only traced IR graph stores module typename and variable name
in `scope` in `node`. This change enables such `scope` info for IR graph
generated by torch script.
Torch script produced IR graphs emit nodes for module object and module
method call. This structured graph is flattened in `function_substition`
pass prior to other ONNX conversion passes.
This PR extends `function_substition` pass to record the module typename
and variable name info in `scope`, while inlining the graph.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Assign ONNXScopeName during function substituion"


Previously only traced IR graph stores module typename and variable name
in `scope` in `node`. This change enables such `scope` info for IR graph
generated by torch script.
Torch script produced IR graphs emit nodes for module object and module
method call. This structured graph is flattened in `function_substition`
pass prior to other ONNX conversion passes.
This PR extends `function_substition` pass to record the module typename
and variable name info in `scope`, while inlining the graph.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Assign ONNXScopeName during function substituion"


Previously only traced IR graph stores module typename and variable name
in `scope` in `node`. This change enables such `scope` info for IR graph
generated by torch script.
Torch script produced IR graphs emit nodes for module object and module
method call. This structured graph is flattened in `function_substition`
pass prior to other ONNX conversion passes.
This PR extends `function_substition` pass to record the module typename
and variable name info in `scope`, while inlining the graph.

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Assign ONNXScopeName during function substituion"


Previously only traced IR graph stores module typename and variable name
in `scope` in `node`. This change enables such `scope` info for IR graph
generated by torch script.
Torch script produced IR graphs emit nodes for module object and module
method call. This structured graph is flattened in `function_substition`
pass prior to other ONNX conversion passes.
This PR extends `function_substition` pass to record the module typename
and variable name info in `scope`, while inlining the graph.

[ghstack-poisoned]

BowenBao commented:
@abock for approval

justinchuby commented on torch/csrc/jit/passes/onnx/function_substitution.cpp:
```diff
@@ -2,11 +2,84 @@
 
+namespace {
+
+const std::string top_module_variable_name = "";
```
k for constant?

BowenBao replied:
It's between this (snake case convention) or k with camel case. I'm not too sure about k with snake case, something like `k_top_module_variable_name` is misleading. Going with k with camel case, and I think we will need to move away from camel case naming conventions for functions as they are too similar.

justinchuby replied:
Yeah I was suggesting kCamelCase name. Moving away from camel case for functions sounds like the right thing to do. We just need to do that consistently at least across the torch onnx namespaces.

justinchuby commented on torch/csrc/jit/passes/onnx/function_substitution.cpp:
```diff
@@ -2,11 +2,84 @@
+  const std::string& method_name = call_node->s(attr::name);
+  if (method_name == "forward") {
+    const auto type = call_node->input(0)->type()->expect<c10::NamedType>();
+    const auto class_name = TidyClassNameFromTorchScript(type->name());
+    const auto variable_name = GetCallNodeVariableName(call_node);
+    const auto scope_name =
```
Perhaps use explicit types for these variables for clarity?

BowenBao replied:
Done for the 3 string cases. I think they are fine either way. It wouldn't hurt readability too much, while there weren't much convenience offered by auto, since I don't expect the function signature and return type to change in the future.
Keeping `type` as is, since its type is explicitly spelled out in `expect` later.

BowenBao committed:
Update on "[ONNX] Assign ONNXScopeName during function substituion"


Previously only traced IR graph stores module typename and variable name
in `scope` in `node`. This change enables such `scope` info for IR graph
generated by torch script.
Torch script produced IR graphs emit nodes for module object and module
method call. This structured graph is flattened in `function_substition`
pass prior to other ONNX conversion passes.
This PR extends `function_substition` pass to record the module typename
and variable name info in `scope`, while inlining the graph.

[ghstack-poisoned]

BowenBao commented:
@abock friendly ping for this one :)

BowenBao commented:
@pytorchbot merge -g
